<a href="https://colab.research.google.com/github/astrovishalthakur/MachineLearning/blob/main/NaturalLanguageProcessing/Basics/TextPreprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### NLP PIPELINE

We will se how to do basic text processing as a nlp pipeline

In [133]:
import requests

In [134]:
url1 = "https://api.themoviedb.org/3/movie/top_rated?api_key=73326928ce817d080bf46bdc0bc675b2&language=en-US&page="

In [135]:
url2 = "https://api.themoviedb.org/3/genre/movie/list?api_key=73326928ce817d080bf46bdc0bc675b2&language=en-US"

In [136]:
details = []

In [137]:
data = requests.get(f'{url1}1')

In [138]:
pages = data.json()["total_pages"]

In [139]:
type(data.json()["total_pages"])

int

In [140]:
type(data.json()["results"])

list

In [141]:
for i in range(1, pages+1):
  data = requests.get(f"{url1}{i}")
  details.extend(data.json()["results"])

In [142]:
len(details)

9720

In [143]:
details[1]

{'adult': False,
 'backdrop_path': '/wPU78OPN4BYEgWYdXyg0phMee64.jpg',
 'genre_ids': [18, 80],
 'id': 278,
 'original_language': 'en',
 'original_title': 'The Shawshank Redemption',
 'overview': 'Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.',
 'popularity': 103.61,
 'poster_path': '/q6y0Go1tsGEsmtFryDOJo3dEmqu.jpg',
 'release_date': '1994-09-23',
 'title': 'The Shawshank Redemption',
 'video': False,
 'vote_average': 8.7,
 'vote_count': 20895}

In [144]:
genres = requests.get(url2)
ids = genres.json()

In [145]:
ids

{'genres': [{'id': 28, 'name': 'Action'},
  {'id': 12, 'name': 'Adventure'},
  {'id': 16, 'name': 'Animation'},
  {'id': 35, 'name': 'Comedy'},
  {'id': 80, 'name': 'Crime'},
  {'id': 99, 'name': 'Documentary'},
  {'id': 18, 'name': 'Drama'},
  {'id': 10751, 'name': 'Family'},
  {'id': 14, 'name': 'Fantasy'},
  {'id': 36, 'name': 'History'},
  {'id': 27, 'name': 'Horror'},
  {'id': 10402, 'name': 'Music'},
  {'id': 9648, 'name': 'Mystery'},
  {'id': 10749, 'name': 'Romance'},
  {'id': 878, 'name': 'Science Fiction'},
  {'id': 10770, 'name': 'TV Movie'},
  {'id': 53, 'name': 'Thriller'},
  {'id': 10752, 'name': 'War'},
  {'id': 37, 'name': 'Western'}]}

In [146]:
new_ids = {}
for items in ids["genres"]:
  new_ids[items["id"]]= items["name"]

In [147]:
new_ids

{12: 'Adventure',
 14: 'Fantasy',
 16: 'Animation',
 18: 'Drama',
 27: 'Horror',
 28: 'Action',
 35: 'Comedy',
 36: 'History',
 37: 'Western',
 53: 'Thriller',
 80: 'Crime',
 99: 'Documentary',
 878: 'Science Fiction',
 9648: 'Mystery',
 10402: 'Music',
 10749: 'Romance',
 10751: 'Family',
 10752: 'War',
 10770: 'TV Movie'}

In [148]:
finalLis = []
dic = {}
for item in details:
  dic = {"title":item["title"],
         "genres": [new_ids[i] for i in item["genre_ids"]],
         "description":item["overview"]}
  finalLis.append(dic)
  dic = {}

In [149]:
finalLis[:5]

[{'description': 'Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.',
  'genres': ['Comedy', 'Drama', 'Romance'],
  'title': 'Dilwale Dulhania Le Jayenge'},
 {'description': 'Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.',
  'genres': ['Drama', 'Crime'],
  'title': 'The Shawshank Redemption'},
 {'description': 'Spanning

In [150]:
import pandas as pd

In [151]:
df = pd.DataFrame(finalLis, columns=["description", "genres", "title"])

In [152]:
df.head()

description  \
0  Raj is a rich, carefree, happy-go-lucky second...   
1  Framed in the 1940s for the double murder of h...   
2  Spanning the years 1945 to 1955, a chronicle o...   
3  A tragic accident lead to Kaori's blindness, b...   
4  The true story of how businessman Oskar Schind...   

                     genres                        title  
0  [Comedy, Drama, Romance]  Dilwale Dulhania Le Jayenge  
1            [Drama, Crime]     The Shawshank Redemption  
2            [Drama, Crime]                The Godfather  
3          [Romance, Drama]               Your Eyes Tell  
4     [Drama, History, War]             Schindler's List

In [153]:
df = df[["title", "genres", "description"]]

In [154]:
df

title  \
0          Dilwale Dulhania Le Jayenge   
1             The Shawshank Redemption   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9715                 Alone in the Dark   
9716                 House of the Dead   
9717              Dragonball Evolution   
9718  Box Office 3D - Il film dei film   
9719        Birdemic: Shock and Terror   

                                                 genres  \
0                              [Comedy, Drama, Romance]   
1                                        [Drama, Crime]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9715                          [Action, Fantasy, Horror]   
9716                         [Horror, Action, Thriller]   
9717  [Action, Adventure, Fantasy, Science Fiction, ...   
9718                                           [Comedy]   
9719       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     Raj is a rich, carefree, happy-go-lucky second...  
1     Framed in the 1940s for the double murder of h...  
2     Spanning the years 1945 to 1955, a chronicle o...  
3     A tragic accident lead to Kaori's blindness, b...  
4     The true story of how businessman Oskar Schind...  
...                                                 ...  
9715  Edward Carnby is a private investigator specia...  
9716  Set on an island off the coast, a techno rave ...  
9717  The young warrior Son Goku sets out on a quest...  
9718  An Italian parody of the biggest U.S. blockbus...  
9719  A platoon of eagles and vultures attacks the r...  

[9720 rows x 3 columns]

# Making everything Lower Case

In [155]:
df.description = df.description.str.lower()

In [156]:
df.description

0       raj is a rich, carefree, happy-go-lucky second...
1       framed in the 1940s for the double murder of h...
2       spanning the years 1945 to 1955, a chronicle o...
3       a tragic accident lead to kaori's blindness, b...
4       the true story of how businessman oskar schind...
                              ...                        
9715    edward carnby is a private investigator specia...
9716    set on an island off the coast, a techno rave ...
9717    the young warrior son goku sets out on a quest...
9718    an italian parody of the biggest u.s. blockbus...
9719    a platoon of eagles and vultures attacks the r...
Name: description, Length: 9720, dtype: object

# Removing HTML Tags

example

In [157]:
import re
def striphtml(data):
    p = re.compile(r'<.*?>')
    return p.sub('', data)

striphtml('<a href="foo.com" class="bar">I Want This <b>text!</b></a>')


'I Want This text!'

In [158]:
k = df.description.apply(lambda x:striphtml(x))

In [159]:
for item1, item2 in zip(k, df.description):
  if item1 != item2:
    print(item2)

In [160]:
dic = dict(k == df.description)

In [161]:
type(dic)

dict

In [162]:
for i in dic:
  if dic[i] == False:
    print(dic[i])

since above code gave no output, there were no html tags in our description

Looks like we don't need to remove html tags

let's make sure with this example that above approach is right to determine whether we want to remove html tags.

In [163]:
sr = pd.Series(["hello", "my name is <p>vishal</p>thakur", "Who <html>are</html>you"])

In [164]:
sr

0                             hello
1    my name is <p>vishal</p>thakur
2           Who <html>are</html>you
dtype: object

In [165]:
kk = sr.apply(lambda x: striphtml(x))

In [166]:
kk

0                      hello
1    my name is vishalthakur
2                 Who areyou
dtype: object

In [167]:
dic2 = dict(kk == sr)

In [168]:
dic2

{0: True, 1: False, 2: False}

In [169]:
for i in dic2:
  if dic2[i] == False:
    print(i)

1
2


as you can see, above code gave output of 1 and 2, but same code didn't gave any output on our description.

# Removing Puntuations

In [170]:
pattern = re.compile(r'[^\w\s]')

In [171]:
re.sub(pattern,"", "hello! my name is ? vishal")

'hello my name is  vishal'

In [172]:
df.description = df.description.apply(lambda x: re.sub(pattern, "", x))

In [173]:
df.description

0       raj is a rich carefree happygolucky second gen...
1       framed in the 1940s for the double murder of h...
2       spanning the years 1945 to 1955 a chronicle of...
3       a tragic accident lead to kaoris blindness but...
4       the true story of how businessman oskar schind...
                              ...                        
9715    edward carnby is a private investigator specia...
9716    set on an island off the coast a techno rave p...
9717    the young warrior son goku sets out on a quest...
9718    an italian parody of the biggest us blockbuste...
9719    a platoon of eagles and vultures attacks the r...
Name: description, Length: 9720, dtype: object

In [174]:
df

title  \
0          Dilwale Dulhania Le Jayenge   
1             The Shawshank Redemption   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9715                 Alone in the Dark   
9716                 House of the Dead   
9717              Dragonball Evolution   
9718  Box Office 3D - Il film dei film   
9719        Birdemic: Shock and Terror   

                                                 genres  \
0                              [Comedy, Drama, Romance]   
1                                        [Drama, Crime]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9715                          [Action, Fantasy, Horror]   
9716                         [Horror, Action, Thriller]   
9717  [Action, Adventure, Fantasy, Science Fiction, ...   
9718                                           [Comedy]   
9719       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     raj is a rich carefree happygolucky second gen...  
1     framed in the 1940s for the double murder of h...  
2     spanning the years 1945 to 1955 a chronicle of...  
3     a tragic accident lead to kaoris blindness but...  
4     the true story of how businessman oskar schind...  
...                                                 ...  
9715  edward carnby is a private investigator specia...  
9716  set on an island off the coast a techno rave p...  
9717  the young warrior son goku sets out on a quest...  
9718  an italian parody of the biggest us blockbuste...  
9719  a platoon of eagles and vultures attacks the r...  

[9720 rows x 3 columns]

# Spelling correction

In [175]:
! pip install textblob

In [176]:
from textblob import TextBlob

In [177]:
str1 = "helo, my nami is vishal Thakur. Who are you"

In [178]:
blb = TextBlob(str1)

In [179]:
blb.correct().string

'held, my name is visual Thakur. Who are you'

looks like it is making some mistakes on non english words. Also, since this is movied descriptions, it will not have many spelling mistakes. So let's not do any spell corrections

# Stop words treatment

In [180]:
! pip install nltk

In [181]:
import nltk

In [182]:
nltk.download("stopwords")
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [183]:
stopwords.words("english")[:10]

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [184]:
len(stopwords.words("english"))

179

looks like we have 179 stop words

We will split the text, then remove stop words.

In [185]:
# let's see how join works
a = ["a", "b", "c", "d"]
" ".join(a)

'abcd'

In [186]:
def remove_stop(text):
  lis = []
  for word in text.split():
    if word not in stopwords.words("english"):
      lis.append(word)
  return " ".join(lis)


as you can see, we are joining every thing back. We will be tokenizing everything again, so we don't need to join it back. We can just return the list.

`return lis`

In [187]:
series = df.description.apply(lambda x: remove_stop(x))

In [188]:
df.description = series

In [189]:
df

title  \
0          Dilwale Dulhania Le Jayenge   
1             The Shawshank Redemption   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9715                 Alone in the Dark   
9716                 House of the Dead   
9717              Dragonball Evolution   
9718  Box Office 3D - Il film dei film   
9719        Birdemic: Shock and Terror   

                                                 genres  \
0                              [Comedy, Drama, Romance]   
1                                        [Drama, Crime]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9715                          [Action, Fantasy, Horror]   
9716                         [Horror, Action, Thriller]   
9717  [Action, Adventure, Fantasy, Science Fiction, ...   
9718                                           [Comedy]   
9719       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     raj rich carefree happygolucky second generati...  
1     framed 1940s double murder wife lover upstandi...  
2     spanning years 1945 1955 chronicle fictional i...  
3     tragic accident lead kaoris blindness clings l...  
4     true story businessman oskar schindler saved t...  
...                                                 ...  
9715  edward carnby private investigator specializin...  
9716  set island coast techno rave party attracts di...  
9717  young warrior son goku sets quest racing time ...  
9718  italian parody biggest us blockbusters gladiat...  
9719  platoon eagles vultures attacks residents smal...  

[9720 rows x 3 columns]

# Tokenization
It's time to tokenize

In [190]:
from nltk.tokenize import word_tokenize
nltk.download("punkt")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [191]:
word_tokenize("Hello my name is vishal thakur.")

['Hello', 'my', 'name', 'is', 'vishal', 'thakur', '.']

In [194]:
df.description.iloc[0]

'raj rich carefree happygolucky second generation nri simran daughter chaudhary baldev singh spite nri strict adherence indian values simran left india married childhood fiancé raj leaves india mission hands claim lady love noses whole family thus begins saga'

Checking on one example

In [193]:
word_tokenize(df.description.iloc[0])

['raj',
 'rich',
 'carefree',
 'happygolucky',
 'second',
 'generation',
 'nri',
 'simran',
 'daughter',
 'chaudhary',
 'baldev',
 'singh',
 'spite',
 'nri',
 'strict',
 'adherence',
 'indian',
 'values',
 'simran',
 'left',
 'india',
 'married',
 'childhood',
 'fiancé',
 'raj',
 'leaves',
 'india',
 'mission',
 'hands',
 'claim',
 'lady',
 'love',
 'noses',
 'whole',
 'family',
 'thus',
 'begins',
 'saga']

Applying to whole description column of dataset

In [197]:
df.description = df.description.apply(lambda x: word_tokenize(x))
df.description

0       [raj, rich, carefree, happygolucky, second, ge...
1       [framed, 1940s, double, murder, wife, lover, u...
2       [spanning, years, 1945, 1955, chronicle, ficti...
3       [tragic, accident, lead, kaoris, blindness, cl...
4       [true, story, businessman, oskar, schindler, s...
                              ...                        
9715    [edward, carnby, private, investigator, specia...
9716    [set, island, coast, techno, rave, party, attr...
9717    [young, warrior, son, goku, sets, quest, racin...
9718    [italian, parody, biggest, us, blockbusters, g...
9719    [platoon, eagles, vultures, attacks, residents...
Name: description, Length: 9720, dtype: object

# Stemming

In [196]:
# import these modules
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

ps = PorterStemmer()

# choose some words to be stemmed
words = ["program", "programs", "programmer", "programming", "programmers"]

for w in words:
	print(w, " : ", ps.stem(w))


program  :  program
programs  :  program
programmer  :  programm
programming  :  program
programmers  :  programm


In [202]:
from nltk.stem import WordNetLemmatizer
wordnet = WordNetLemmatizer()

In [205]:
nltk.download("wordnet")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

# Learning Formatted Output

That's how we align things.

In [211]:

# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed on the left side
print(f"{'Left Aligned Text' : <20}")

Left Aligned Text   


In [212]:

# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed on the right side
print(f"{'Right Aligned Text' : >20}")

  Right Aligned Text


In [214]:
# here 20 spaces are reserved for the 
# particular output string. And the string
# is printed in the middle
print(f"{'Centered' : ^10}")

 Centered 


In [215]:
# assigning strings to the variables
left_alignment = "Left Text"
center_alignment = "Centered Text"
right_alignment = "Right Text"

# printing out aligned text
print(f"{left_alignment : <20}{center_alignment : ^15}{right_alignment : >20}")


Left Text            Centered Text           Right Text


In [210]:

# assigning list values to the variables
names = ['Raj', 'Shivam', 'Shreeya', 'Kartik']
marks = [7, 9, 8, 5]
div = ['A', 'A', 'C', 'B']
id = [21, 52, 27, 38]


print(f"{'Name' : <10}{'Marks' : ^10}{'Division' : ^10}{'ID' : >5}")
  
# printing values of variables in Aligned manner
for i in range(0, 4):
    print(f"{names[i] : <10}{marks[i] : ^10}{div[i] : ^10}{id[i] : >5}")

Name        Marks    Division    ID
Raj           7         A        21
Shivam        9         A        52
Shreeya       8         C        27
Kartik        5         B        38


In [226]:
print(f"{'Original < 20'}{'Stemmed':^20}{'Lemmatized':^20}")
for w in df.description[6]:
  print(f"{w:<20}{ps.stem(w):<20}{wordnet.lemmatize(w):<20}")

Original < 20      Stemmed            Lemmatized     
continuing          continu             continuing          
saga                saga                saga                
corleone            corleon             corleone            
crime               crime               crime               
family              famili              family              
young               young               young               
vito                vito                vito                
corleone            corleon             corleone            
grows               grow                grows               
sicily              sicili              sicily              
1910s               1910                1910s               
new                 new                 new                 
york                york                york                
1950s               1950                1950s               
michael             michael             michael             
corleone            corleon    

looks like both are giving good results, but we will manually check for some words

In [228]:
a = ["continue", "continued", "countinuing", "countinous", "continuing"]

for w in a:
  print(w, "  :  ", ps.stem(w), "  :  ", wordnet.lemmatize(w))

continue   :   continu   :   continue
continued   :   continu   :   continued
countinuing   :   countinu   :   countinuing
countinous   :   countin   :   countinous
continuing   :   continu   :   continuing


since lemmatization gives so many version above words, we will go with stemming

In [232]:
des_stemmed = df.description.apply(lambda x: [ps.stem(i) for i in x])

In [233]:
des_stemmed

0       [raj, rich, carefre, happygolucki, second, gen...
1       [frame, 1940, doubl, murder, wife, lover, upst...
2       [span, year, 1945, 1955, chronicl, fiction, it...
3       [tragic, accid, lead, kaori, blind, cling, lif...
4       [true, stori, businessman, oskar, schindler, s...
                              ...                        
9715    [edward, carnbi, privat, investig, special, un...
9716    [set, island, coast, techno, rave, parti, attr...
9717    [young, warrior, son, goku, set, quest, race, ...
9718    [italian, parodi, biggest, us, blockbust, glad...
9719    [platoon, eagl, vultur, attack, resid, small, ...
Name: description, Length: 9720, dtype: object

In [234]:
df.description = des_stemmed

# And here we have our dataset, ready to be modelled upon.

In [235]:
df

title  \
0          Dilwale Dulhania Le Jayenge   
1             The Shawshank Redemption   
2                        The Godfather   
3                       Your Eyes Tell   
4                     Schindler's List   
...                                ...   
9715                 Alone in the Dark   
9716                 House of the Dead   
9717              Dragonball Evolution   
9718  Box Office 3D - Il film dei film   
9719        Birdemic: Shock and Terror   

                                                 genres  \
0                              [Comedy, Drama, Romance]   
1                                        [Drama, Crime]   
2                                        [Drama, Crime]   
3                                      [Romance, Drama]   
4                                 [Drama, History, War]   
...                                                 ...   
9715                          [Action, Fantasy, Horror]   
9716                         [Horror, Action, Thriller]   
9717  [Action, Adventure, Fantasy, Science Fiction, ...   
9718                                           [Comedy]   
9719       [Science Fiction, Horror, Romance, Thriller]   

                                            description  
0     [raj, rich, carefre, happygolucki, second, gen...  
1     [frame, 1940, doubl, murder, wife, lover, upst...  
2     [span, year, 1945, 1955, chronicl, fiction, it...  
3     [tragic, accid, lead, kaori, blind, cling, lif...  
4     [true, stori, businessman, oskar, schindler, s...  
...                                                 ...  
9715  [edward, carnbi, privat, investig, special, un...  
9716  [set, island, coast, techno, rave, parti, attr...  
9717  [young, warrior, son, goku, set, quest, race, ...  
9718  [italian, parodi, biggest, us, blockbust, glad...  
9719  [platoon, eagl, vultur, attack, resid, small, ...  

[9720 rows x 3 columns]